cmake/kokkos_tribits.cmake

```diff
@@ -0,0 +1,420 @@
+  SET(EXE_NAME ${PACKAGE_NAME}_${ROOT_NAME})
+  KOKKOS_ADD_TEST_EXECUTABLE(${EXE_NAME}
+    SOURCES ${PARSE_SOURCES}
+  )
+  KOKKOS_ADD_TEST(NAME ${ROOT_NAME} 
+    EXE ${EXE_NAME}
+    FAIL_REGULAR_EXPRESSION "  FAILED  "
+  )
+ENDIF()
+ENDFUNCTION()
+
+MACRO(KOKKOS_SETUP_BUILD_ENVIRONMENT)
+ IF (NOT KOKKOS_HAS_TRILINOS)
+  #------------ COMPILER AND FEATURE CHECKS ------------------------------------
+  INCLUDE(${KOKKOS_SRC_PATH}/cmake/kokkos_functions.cmake)
+
+  #------------ GET OPTIONS AND KOKKOS_SETTINGS --------------------------------
+  # ADD Kokkos' modules to CMake's module path.
+  SET(CMAKE_MODULE_PATH ${CMAKE_MODULE_PATH} "${Kokkos_SOURCE_DIR}/cmake/Modules/")
+
+  INCLUDE(${KOKKOS_SRC_PATH}/cmake/kokkos_enable_options.cmake)
+  INCLUDE(${KOKKOS_SRC_PATH}/cmake/kokkos_cxx.cmake)
+  INCLUDE(${KOKKOS_SRC_PATH}/cmake/kokkos_tpls.cmake)
+  INCLUDE(${KOKKOS_SRC_PATH}/cmake/kokkos_arch.cmake)
+ ENDIF()
+ENDMACRO(KOKKOS_SETUP_BUILD_ENVIRONMENT)
+
+MACRO(KOKKOS_ADD_TEST_EXECUTABLE EXE_NAME)
+  CMAKE_PARSE_ARGUMENTS(PARSE 
+    ""
+    ""
+    "SOURCES"
+    ${ARGN})
+  KOKKOS_ADD_EXECUTABLE(${EXE_NAME}
+    SOURCES ${PARSE_SOURCES}
+    ${PARSE_UNPARSED_ARGUMENTS}
+    TESTONLYLIBS kokkos_gtest
+  )
+  IF (NOT KOKKOS_HAS_TRILINOS)
+    TARGET_LINK_LIBRARIES(${EXE_NAME} kokkos_gtest)
+    ADD_DEPENDENCIES(check ${EXE_NAME})
+  ENDIF()
+ENDMACRO(KOKKOS_ADD_TEST_EXECUTABLE)
+
+MACRO(KOKKOS_PACKAGE_POSTPROCESS)
+  if (KOKKOS_HAS_TRILINOS)
+    TRIBITS_PACKAGE_POSTPROCESS()
+  endif()
+ENDMACRO(KOKKOS_PACKAGE_POSTPROCESS)
+
+MACRO(KOKKOS_MAKE_LIBKOKKOS)
+  ADD_LIBRARY(kokkos ${KOKKOS_SOURCE_DIR}/core/src/dummy.cpp)
+  TARGET_LINK_LIBRARIES(kokkos PUBLIC kokkoscore kokkoscontainers)
+  TARGET_LINK_LIBRARIES(kokkos PUBLIC kokkosalgorithms)
+  KOKKOS_INTERNAL_ADD_LIBRARY_INSTALL(kokkos)
+ENDMACRO()
+
+FUNCTION(KOKKOS_INTERNAL_ADD_LIBRARY LIBRARY_NAME)
+  CMAKE_PARSE_ARGUMENTS(PARSE 
+    "STATIC;SHARED"
+    ""
+    "HEADERS;SOURCES"
+    ${ARGN})
+
+  IF(PARSE_HEADERS)
+    LIST(REMOVE_DUPLICATES PARSE_HEADERS)
+  ENDIF()
+  IF(PARSE_SOURCES)
+    LIST(REMOVE_DUPLICATES PARSE_SOURCES)
+  ENDIF()
+
+  ADD_LIBRARY(
+    ${LIBRARY_NAME}
+    ${PARSE_HEADERS}
+    ${PARSE_SOURCES}
+  )
+
+  TARGET_COMPILE_OPTIONS(
+    ${LIBRARY_NAME}
+    PUBLIC $<$<COMPILE_LANGUAGE:CXX>:${KOKKOS_COMPILE_OPTIONS}>
+  )
+
+  IF(${CMAKE_VERSION} VERSION_GREATER "3.13" OR ${CMAKE_VERSION} VERSION_EQUAL "3.13")
+    TARGET_LINK_OPTIONS(
+      ${LIBRARY_NAME}
+      PUBLIC ${KOKKOS_LD_FLAGS}
+    )
+  ELSE()
+    #WELl, this is annoying - I am going to need to hack this for Visual Studio
+    TARGET_LINK_LIBRARIES(
+      ${LIBRARY_NAME} PUBLIC ${KOKKOS_LINK_OPTIONS}
+    )
+  ENDIF()
+
+  IF (KOKKOS_ENABLE_CUDA)
+    TARGET_COMPILE_OPTIONS(
+      ${LIBRARY_NAME}
+      PUBLIC ${KOKKOS_CUDA_OPTIONS}>
+    )
+    SET(NODEDUP_CUDAFE_OPTIONS)
+    FOREACH(OPT ${NODEDEUP_CUDAFE_OPTIONS})
+      LIST(APPEND NODEDUP_CUDAFE_OPTIONS "SHELL: -Xcudafe ${OPT}") 
```

Comment:
```suggestion
      LIST(APPEND NODEDUP_CUDAFE_OPTIONS -Xcudafe ${OPT}) 
```

Reply:
Is this breaking your build? If you don't have the SHELL part, then CMake will deduplicate all the -Xcudafe flags, which you don't want.

Reply:
Yes, it is breaking my build. I get
```
/tmp/kokkos/bin/nvcc_wrapper   -I/tmp/kokkos/build_cmake/build/core/src -I/tmp/kokkos/core/src -I/tmp/kokkos/build_cmake/build  -mavx "SHELL: -Xcompiler -fopenmp" -std=c++11 -std=gnu++11 -o CMakeFiles/kokkoscore.dir/impl/Kokkos_CPUDiscovery.cpp.o -c /tmp/kokkos/core/src/impl/Kokkos_CPUDiscovery.cpp
g++: error: SHELL:: No such file or directory
```

Reply:
What version of cmake are you using? Looks like I had my versions wrong. This didn't get added until 3.12, but I set minimum to 3.10.

Reply:
@crtrott: Do we have any scenarios where we pass more than one -Xcudafe flag to the compiler? The only one I see right now is `--diag_suppress=esa_on_defaulted_function_ignored`. If we need to support mort than one `-Xcudafe` we will need to bump minimum to 3.12.

Reply:
I was using `CMake 3.10.2`.